src/hooks/useAsync.ts

```diff
@@ -46,11 +50,39 @@ const useAsync = <Result, Error>(
   }, [asyncFn]);
 
   return result;
-
-  // Loading is false but result hasn't been set;
 };
 
-export default useAsync;
+export type UseAsyncOnDemandReturnType<Result, Error = unknown> = [
+  () => void,
+  UseAsyncReturnType<Result, Error>
+];
+
+export const useAsyncOnDemand = <Args extends unknown[], Result, Error>(
+  asyncFn: (...args: Args) => Promise<Result>
+): UseAsyncOnDemandReturnType<Result, Error> => {
+  const [result, setResult] = useState<UseAsyncReturnType<Result, Error>>({
+    loading: false,
+    result: null,
+    error: null,
+  });
+
+  const funcOnDemand = useCallback(
+    (...args: Args) => {
+      setResult({ result: null, error: null, loading: true });
```

Comment:
I love that this state reset means we can call our callback as many times as we want and it'll work as expected.